src/v/kafka/server/snc_quota_manager.cc

```diff
@@ -309,9 +310,95 @@ void snc_quota_manager::quota_balancer() {
-    return ss::make_ready_future<>();
+    vlog(klog.trace, "qb - Step");
+
+    // determine the borrowers and whether any balancing is needed now
+    const auto now = clock::now();
+    const std::vector<borrower_t> borrowers = co_await container().map(
+      [now](snc_quota_manager& qm) -> borrower_t {
+          qm.refill_buckets(now);
+          return {
+            .shard_id = ss::this_shard_id(),
+            .is_borrower = to_each(
+              [](const quota_t d) -> shard_count_t { return d > 0 ? 1u : 0u; },
+              qm.get_deficiency())};
+      });
+    vlog(klog.trace, "qb - Borrowers: {}", borrowers);
+
+    const auto borrowers_count = std::accumulate(
+      borrowers.cbegin(),
+      borrowers.cend(),
+      inoutpair<shard_count_t>{0u, 0u},
+      [](const inoutpair<shard_count_t>& s, const borrower_t& b) {
+          return to_each(std::plus{}, s, b.is_borrower);
+      });
+    vlog(klog.trace, "qb - Borrowers count: {}", borrowers_count);
+    if (is_zero(borrowers_count)) {
+        co_return;
+    }
+
+    // collect quota from lenders
+    const inoutpair<quota_t> collected = co_await container().map_reduce0(
+      [borrowers_count](snc_quota_manager& qm) {
+          const auto to_collect = to_each(
+            [](const quota_t surplus, const shard_count_t borrowers_count) {
+                return borrowers_count > 0 ? surplus / 2 : 0;
+            },
+            qm.get_surplus(),
+            borrowers_count);
+          qm.adjust_quota(to_each(std::negate{}, to_collect));
+          return to_collect;
+      },
+      inoutpair<quota_t>{0, 0},
+      [](const inoutpair<quota_t>& s, const inoutpair<quota_t>& v) {
+          return to_each(std::plus{}, s, v);
+      });
+    vlog(klog.trace, "qb - Collected: {}", collected);
+
+    // dispense the collected amount among the borrowers
+    auto split = to_each(
+      [](const quota_t collected, const shard_count_t borrowers_count) {
+          if (borrowers_count == 0) {
+              vassert(
+                collected == 0, "Quota collected when borrowers are absent");
+              return std::div(collected, shard_count_t{1});
+          }
+          return std::div(collected, borrowers_count);
+      },
+      collected,
+      borrowers_count);
+    for (const borrower_t& b : borrowers) {
+        if (b.is_borrower.in == 0 && b.is_borrower.out == 0) {
+            continue;
+        }
+        const inoutpair<quota_t> share = to_each(
+          [](auto& split, const shard_count_t is_borrower) -> quota_t {
+              if (is_borrower == 0) {
+                  return 0;
+              }
+              if (split.rem == 0) {
+                  return split.quot;
+              }
+              --split.rem;
+              return split.quot + 1;
+          },
+          split,
+          b.is_borrower);
+        // TBD: make the invocation parallel
+        co_await container().invoke_on(
+          b.shard_id,
```

Comment:
are there some distinct components here that could be factored out into separate functions and explain indivdiually? ok if there is not.